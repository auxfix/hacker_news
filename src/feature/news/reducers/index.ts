export { newsState, default as news } from './news';
