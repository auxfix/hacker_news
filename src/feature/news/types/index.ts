export interface hackerstory {
  authorid: string;
  id: number;
  img?: string;
  karma?: number;
  score: number;
  time: number;
  title: string;
  url: string;
}

export interface user {
  about: string;
  created: number;
  id: string;
  karma: number;
}
