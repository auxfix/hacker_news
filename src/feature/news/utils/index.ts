export function getRundomItemsFromArray(array: Array<number>, numberOfRundomElements: number): Array<number>{
    return array.slice().sort(() => .5 - Math.random()).slice(0,numberOfRundomElements);
}; 

export function getRundomArray(arrayLength: number): Array<number> {
    return Array.from(Array(arrayLength).keys()).sort(() => .5 - Math.random());
}


