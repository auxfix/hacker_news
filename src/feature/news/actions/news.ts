import { createAction } from '@reduxjs/toolkit';
import { ActionTypes } from 'literals';
import { actionPayload } from 'modules/helpers';
import { HackerStory } from 'types';

export const getNews = createAction(ActionTypes.GET_NEWS_REQUEST);
export const getNewsSuccess = createAction(ActionTypes.GET_NEWS_SUCCESS, (payload: HackerStory[]) =>
  actionPayload(payload),
);
export const getNewsFailure = createAction(ActionTypes.GET_NEWS_FAILURE);
