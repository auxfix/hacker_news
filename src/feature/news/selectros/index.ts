import { createSelector } from '@reduxjs/toolkit';

import { RootState } from 'types';

export const selectHackerNews = createSelector(
    (state: RootState) => state.news.news,
    news => news,
);

export const selectNewsAreLoading = createSelector(
    (state: RootState) => state.news.isLoading,
    isLoading => isLoading,
);

