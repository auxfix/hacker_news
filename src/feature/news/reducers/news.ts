import { createReducer } from '@reduxjs/toolkit';
import { getNews, getNewsFailure, getNewsSuccess } from 'feature/news/actions';
import { NewsState } from 'types';

export const newsState: NewsState = {
  news: [],
  isLoading: false,
};

export default {
  news: createReducer<NewsState>(newsState, builder => {
    builder
      .addCase(getNews, draft => {
        draft.isLoading = true;
      })
      .addCase(getNewsSuccess, (draft, { payload }) => {
        draft.news = payload;
        draft.isLoading = false;
      })
      .addCase(getNewsFailure, draft => {
        draft.isLoading = false;
      });
  }),
};
