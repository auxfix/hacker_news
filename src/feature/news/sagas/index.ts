import { getNewsFailure, getNewsSuccess, showAlert } from 'actions';
import rundom_images from 'feature/news/assets/img/rundom_huckers_images';
import HackerApi from 'feature/news/services/api';
import { HackerStoryDTO, UserDTO } from 'feature/news/types/dto';
import { HackerStory_Dto_to_Internal } from 'feature/news/types/typeConverters';
import { getRundomArray, getRundomItemsFromArray } from 'feature/news/utils';
import { ActionTypes } from 'literals';
import { all, put, takeLatest } from 'redux-saga/effects';

const NUMBER_OF_RUNDOM_NEWS_TO_SHOW = 10;

export function* getNews() {
  const hacker_api = new HackerApi();

  try {
    const topNews: Array<number> = yield hacker_api.getTopStories();
    const rundom_10_news = getRundomItemsFromArray(topNews, NUMBER_OF_RUNDOM_NEWS_TO_SHOW);
    const rundom_image_indexes = getRundomArray(NUMBER_OF_RUNDOM_NEWS_TO_SHOW);
    const news: Array<HackerStoryDTO> = [];
    const users: Array<UserDTO> = [];

    for (const id of rundom_10_news) {
      const localNews: HackerStoryDTO = yield hacker_api.getHackerStory(id);

      if (localNews) {
        news.push(localNews);
      }
    }

    const usersIds: Array<string> = news.reduce((acc, item) => {
      if (!acc.includes(item.by)) {
        acc.push(item.by);
      }

      return acc;
    }, [] as Array<string>);

    for (const id of usersIds) {
      const localUser: UserDTO = yield hacker_api.getUser(id);

      if (localUser) {
        users.push(localUser);
      }
    }

    const newsWithAllFiledsPopulated = news.map((newsItem, index) => {
      const newsItemAuthor = users.find(user => user.id === newsItem.by);

      const internalNewsItem = HackerStory_Dto_to_Internal(newsItem);

      internalNewsItem.img = rundom_images[rundom_image_indexes[index]];

      if (newsItemAuthor) {
        internalNewsItem.karma = newsItemAuthor?.karma;
      }

      return internalNewsItem;
    });

    yield put(getNewsSuccess(newsWithAllFiledsPopulated));
  } catch {
    yield put(getNewsFailure());
    yield put(
      showAlert('Error getting news from the internet. Please try again later.', {
        variant: 'danger',
        timeout: 10,
      }),
    );
  }
}

export default function* root() {
  yield all([takeLatest(ActionTypes.GET_NEWS_REQUEST, getNews)]);
}
