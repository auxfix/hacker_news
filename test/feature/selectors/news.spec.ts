import { selectHackerNews, selectNewsAreLoading } from 'feature/news/selectros';
import { configStore } from 'store';

describe('selectors/news', () => {
  const { store } = configStore();

  it('should return the hacker news', () => {
    expect(selectHackerNews(store.getState())).toMatchSnapshot();
  });

  it('should return loading status', () => {
    expect(selectNewsAreLoading(store.getState())).toMatchSnapshot();
  });
});
