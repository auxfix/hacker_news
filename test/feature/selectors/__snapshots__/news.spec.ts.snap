// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/news should return loading status 1`] = `false`;

exports[`selectors/news should return the hacker news 1`] = `[]`;
